README.md
# Fruit Classification with Tensorflow

## Goal:
 - To implement a Convolutional Neural Network in Tensorflow which can accurately disntinguish fruits from each other
 - To undergo an incremental devleopment cycle in building the model, starting with a few classes and building upwards

## Results:
 - Final model trained to classify 40 fruits. Successful with 92.47% Accuracy
 - Structure of model and statistics about it's success during each step of the design process are in changelog.txt

Data set used: fruits-360 dataset from Horea Muresan, Mihai Oltean, Fruit recognition from images using deep learning, Technical Report, Babes-Bolyai University, 2017




changelog.txt
run 7
	Epochs: 30
	Learning Rate: 0.75
	Batch size: 59
	Classes Learned: 10
	Optimizer: Stochastic Gradient Descent

	Structure: 
		Conv (filter = [20 20 3 10], stride = 1, pool = 2) -> 
		Conv (filter = [5 5 10 20], stride=1, pool=2) -> 
		Dense [1000] -> 
		Dense [100] -> 
		Softmax

	Activation Function: Sigmoid
	Loss Function: Cross Entropy
	Regularization: None
	Dropout: 0.0

	Final Loss: 1.547
	Train Time: 2h 26m
	Final Validation Accuracy: 79.13%
	Final Training Accuracy: Not Collected

run 8	
	Epochs: 20
	Learning Rate: 0.75
	Batch size: 59
	Classes Learned: 10
	Optimizer: Momentum (momentum = 0.5)

	Structure: 
		Conv (filter = [20 20 3 10], stride = 1, pool = 2) -> 
		Conv (filter = [5 5 10 20], stride=1, pool=2) -> 
		Dense [1000] -> 
		Dense [100] -> 
		Softmax

	Activation Function: Sigmoid
	Loss Function: Cross Entropy
	Regularization: None
	Dropout: 0.0

	Final Loss: 1.477
	Train Time: 1hr 14m
	Final Validation Accuracy: 89.94%
	Final Training Accuracy: 89.21%

run 9
	Epochs: 20
	Learning Rate: 0.75
	Batch size: 59
	Classes Learned: 10
	Optimizer: Momentum (momentum = 0.5)

	Structure: 
		Conv (filter = [20 20 3 10], stride = 1, pool = 2) -> 
		Conv (filter = [5 5 10 20], stride=1, pool=2) -> 
		Dense [1000] -> 
		Dense [100] -> 
		Softmax

	Activation Function: Sigmoid
	Loss Function: Cross Entropy
	Regularization: None
	Dropout: 0.5

	Final Loss: 2.03
	Train Time: 1hr 26m
	Final Validation Accuracy: 10.06% 
	Final Training Accuracy: NA
	Notes: Stopped after 10 epochs. Model was not training

run 10
	Epochs: 20
	Learning Rate: 0.75
	Batch size: 59
	Classes Learned: 10
	Optimizer: Momentum (momentum = 0.75)

	Structure: 
		Conv (filter = [20 20 3 10], stride = 1, pool = 4) -> 
		Conv (filter = [5 5 10 20], stride=1, pool=2) -> 
		Dense [1000] -> 
		Dense [100] -> 
		Softmax

	Activation Function: Sigmoid
	Loss Function: Cross Entropy
	Regularization: None
	Dropout: 1

	Final Loss: 2.302
	Train Time: 1hr 8m
	Final Validation Accuracy: 8.69% 
	Final Training Accuracy: 10.48%

run 11
	Epochs: 20
	Learning Rate: 0.001
	Batch size: 59
	Classes Learned: 10
	Optimizer: Adam (beta1=0.9, beta2=0.999))

	Structure: 
		Conv (filter = [20 20 3 10], stride = 1, pool = 4) -> 
		Conv (filter = [5 5 10 20], stride=1, pool=2) -> 
		Dense [1000] -> 
		Dense [100] -> 
		Softmax

	Activation Function: Sigmoid
	Loss Function: Cross Entropy
	Regularization: None
	Dropout: 1

	Final Loss: 1.523
	Train Time: 1hr 25min
	Final Validation Accuracy: 98.09%
	Final Training Accuracy: 98.00%

run 12
	Epochs: 5
	Learning Rate: 0.001
	Batch size: 59
	Classes Learned: 10
	Optimizer: Adam (beta1=0.9, beta2=0.999))

	Structure: 
		Conv (filter = [20 20 3 10], stride = 1, pool = 4) -> 
		Conv (filter = [5 5 10 20], stride=1, pool=2) -> 
		Dense [1000] -> 
		Dense [100] -> 
		Softmax

	Activation Function: Leaky ReLU (alpha = 0.2)
	Loss Function: Cross Entropy
	Regularization: None
	Dropout: 1

	Final Loss: 8.552e-5 
	Train Time: 22m
	Final Validation Accuracy: 98.31%
	Final Training Accuracy: 98.74%

run 13
	Epochs: 20
	Learning Rate: 0.001
	Batch size: 59
	Classes Learned: 40
	Optimizer: Adam (beta1=0.9, beta2=0.999))

	Structure: 
		Conv (filter = [20 20 3 10], stride = 1, pool = 4) -> 
		Conv (filter = [5 5 10 20], stride=1, pool=2) -> 
		Dense [1000] -> 
		Dense [100] -> 
		Softmax

	Activation Function: Leaky ReLU (alpha = 0.2)
	Loss Function: Cross Entropy
	Regularization: None
	Dropout: 1

	Final Loss: 4.1983e-4 
	Train Time: 20h 22m
	Final Validation Accuracy: 91.95%
	Final Training Accuracy: 92.47%


dataset.py
import cv2
import os
import numpy as np
import random
import pickle
import h5py as h5

TRAINING_DIR = "fruits-360/Training/"
TEST_DIR = "fruits-360/Test/"
CLASSES = ["Banana", "Lemon", "Mango", "Orange", "Pineapple", "Guava", "Carambula", "Cocos", "Granadilla", "Kiwi", "Apple Braeburn", "Apple Granny Smith", "Apple Red 1", "Apricot", "Avocado", "Cactus Fruit", "Cherry 1", "Cherry 2", "Dates", "Cantaloupe 1", "Cantaloupe 2", "Limes", "Lychee", "Plum", "Raspberry", "Tamarillo", "Tangelo", "Pear", "Kumquats", "Mandarine", "Pepino", "Physalis", "Pineapple Mini", "Salak", "Papaya", "Quince", "Kaki", "Maracuja", "Mulberry", "Pomegranate"]

def image_to_array(imagePath):
    #Convert an image path to an array that can be used for classification
    return cv2.imread(imagePath)

def get_label(label, one_hot=False):
    if not one_hot:
        return CLASSES.index(label)
    vector = np.zeros(len(CLASSES))
    vector[(CLASSES.index(label))] = 1
    return vector

def get_training_data():
    try:
        training_file = h5.File(TRAINING_DIR+'fruit_data.hdf5', 'r')
        return training_file["Training"]["images"], training_file["Training"]["labels"]
    except:
        print("Constructing training data from scratch")
        paths_and_labels = [] 
        for subdir, dirs, files in os.walk(TRAINING_DIR):
            for file in files:
                label = os.path.basename(subdir)
                if label in CLASSES:
                    path = os.path.join(subdir, file)
                    paths_and_labels.append((path, get_label(label)))

        random.shuffle(paths_and_labels)
        split_index = len(paths_and_labels) // 5

        training_half = paths_and_labels[split_index:]
        validation_half = paths_and_labels[:split_index]

        training_file = h5.File(TRAINING_DIR+'fruit_data.hdf5', 'w')
        
        training_images, training_labels = format_data(training_half, training_file, "Training")
        validation_images, validation_labels = format_data(validation_half, training_file, "Validation")
        
        return training_images, training_labels
    #try:
     #   training_pickle = open(os.path.join(TRAINING_DIR, "training.pkl"), 'rb')
      #  images, labels = pickle.load(training_pickle)
       # print("Loaded training_data from pickle")
        #training_pickle.close()

        #combo = list(zip(images, labels))
        #random.shuffle(combo)
        #images, labels = zip(*combo)

        #return images, labels

def get_test_data(batch_size):
    data = []
    for subdir, dirs, files in os.walk(TRAINING_DIR):
        for file in files:
            label = os.path.basename(subdir)
            if label in CLASSES:
                path = os.path.join(subdir, file)
                data.append((image_to_array(path), get_label(label)))

    test_images, test_labels = format_data(data)
    iteration = 1
    while batch_size * iteration < len(data):
        iteration += 1

    return test_images[:batch_size * (iteration - 1)], test_labels[:batch_size * (iteration - 1)]

def get_validation_data(batch_size): 
    training_file = h5.File(TRAINING_DIR+'fruit_data.hdf5', 'r')
    images, labels = training_file["Validation"]["images"], training_file["Validation"]["labels"]
    iteration = 1
    while batch_size * iteration < len(images):
        iteration += 1

    return images[:batch_size * (iteration - 1)], labels[:batch_size * (iteration - 1)]

def format_data(data, outputFile, name):
    shape = (len(data), 100, 100, 3)
    group = outputFile.create_group(name)

    imageSet = group.create_dataset("images", shape, dtype=np.float64)
    labelSet = group.create_dataset("labels", (len(data),), dtype="i")

    random.shuffle(data)

    for index, image_label_pair in enumerate(data):
        imageSet[index,:,:,:] = image_to_array(image_label_pair[0])
        labelSet[index] = image_label_pair[1]
    return imageSet, labelSet


classifier.py
import tensorflow as tf
import dataset
import random

import sys
from tensorflow.python import debug as tf_debug

NUM_CLASSES = len(dataset.CLASSES)
IMAGE_SIZE = 100

def placeholder_inputs(batch_size):
    #Generates placeholders for the inputs
    images_placeholder = tf.placeholder(tf.float32, shape=(batch_size, IMAGE_SIZE, IMAGE_SIZE, 3), name="images")
    labels_placeholder = tf.placeholder(tf.int32, shape=(batch_size,), name="labels")
    return images_placeholder, labels_placeholder

def get_weights(shape):
    initial = tf.truncated_normal(shape, stddev = 0.1)
    return tf.Variable(initial, name="Weights")

def get_biases(shape):
    initial = tf.zeros(shape)
    return tf.Variable(initial, name="Biases")

def conv_layer(input, filter_size, name, strides = [1, 1, 1, 1], pool_size=2, padding="SAME"):
    with tf.name_scope(name):
        weights = get_weights(filter_size)
        biases = get_biases([filter_size[3]])

        #Convolve the image
        #conv = tf.nn.sigmoid(tf.nn.conv2d(input, weights, strides, padding))
        conv = tf.nn.leaky_relu(tf.nn.conv2d(input, weights, strides, padding))
        #Pools the convoluted layer. Arguments are input, pool_size, strides, padding
        pool = tf.nn.max_pool(conv, [1, pool_size, pool_size, 1], [1, pool_size, pool_size, 1], padding)

        return pool

def dense_layer(inputs, input_size, output_size, name):
    with tf.name_scope(name):
        weights = get_weights([input_size, output_size])
        biases = get_biases([output_size])

        h = tf.matmul(inputs, weights) + biases


        #return tf.nn.sigmoid(h)
        return tf.nn.leaky_relu(h)

def define_model(images):
    #Output shape = (batch_size, 25, 25, 10)
    conv1 = conv_layer(images, [20, 20, 3, 10], "Conv1", pool_size=4)
    #Output shape = (batch_size, 13, 13, 20)
    conv2 = conv_layer(conv1, [5, 5, 10, 20], "Conv2")
    conv2 = tf.reshape(conv2, [-1, 13 * 13 * 20])

    dense1 = dense_layer(conv2, 13 * 13 * 20, 1000, "Dense1")
    dense2 = dense_layer(dense1, 1000, 100, "Dense2")
    logits = dense_layer(dense2, 100, NUM_CLASSES, "Softmax_Linear")

    return logits

def define_loss(logits, labels):
    labels = tf.to_int64(labels)

    cross_entropy = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=labels, logits=logits, name="xentropy")
    loss = tf.reduce_mean(cross_entropy, name="xentropy_mean")
    return loss

def training(loss, learning_rate, momentum = 0.75, beta1=0.9, beta2=0.999):
    tf.summary.scalar('loss', loss)
    #optimizer = tf.train.GradientDescentOptimizer(learning_rate)
    #optimizer = tf.train.MomentumOptimizer(learning_rate, momentum)
    optimizer = tf.train.AdamOptimizer(learning_rate, beta1=beta1, beta2=beta2)
    train_op = optimizer.minimize(loss)
    return train_op

def evaluation(logits, labels):
    correct = tf.nn.in_top_k(logits, labels, 1)
    return tf.reduce_sum(tf.cast(correct, tf.int32))

def do_evaluation(session, eval_correct, testDataset, batch_size, images_placeholder, labels_placeholder):
    training_data, training_labels = testDataset
    batches = [(training_data[i:i+batch_size], training_labels[i:i+batch_size]) for i in range(0, len(training_data), batch_size)]
    totalCorrect = 0
    for batch in batches:
        image_data = batch[0]
        label_data = batch[1]

        feed_dict = {images_placeholder: image_data, labels_placeholder: label_data}
        totalCorrect += session.run(eval_correct, feed_dict=feed_dict)

    precision = float(totalCorrect) / len(training_data)
    print('Num examples: %d  Num correct: %d  Precision @ 1: %0.04f' % (len(training_data), totalCorrect, precision))
    sys.stdout.flush()
    return precision > 0.95

def run_training(batch_size, learning_rate, epochs, run_number):
    with tf.Graph().as_default():
        images_placeholder, labels_placeholder= placeholder_inputs(batch_size)

        logits = define_model(images_placeholder)
        lossFunction = define_loss(logits, labels_placeholder)
        train_op = training(lossFunction, learning_rate)

        eval_correct = evaluation(logits, labels_placeholder)

 #       summary = tf.summary.merge_all()
        saver = tf.train.Saver()
        init = tf.global_variables_initializer()

        with tf.Session() as session:
            #session = tf_debug.LocalCLIDebugWrapperSession(session)
            
            logdir = "log/"+str(run_number)
 #           summary_writer = tf.summary.FileWriter(logdir, session.graph)
            session.run(init)

            for step in range(epochs):
                training_data, training_labels = dataset.get_training_data()
                batches = [(training_data[i:i+batch_size], training_labels[i:i+batch_size]) for i in range(0, len(training_data), batch_size)]
                epochLoss = 0
                for batch in batches:
                    image_data = batch[0]
                    label_data = batch[1]

                    feed_dict = {images_placeholder: image_data, labels_placeholder: label_data}
                    activations, loss_value = session.run([train_op, lossFunction], feed_dict=feed_dict)
                    epochLoss += loss_value
                if step % 2 == 0:
                    #print('Step %d: loss = %.2f' % (step, epochLoss))
                    print('Step %d: loss = %.2f' % (step, loss_value))
                    sys.stdout.flush()
#                    summary_str = session.run(summary, feed_dict=feed_dict)
#                    summary_writer.add_summary(summary_str, step)
#                    summary_writer.flush()


                early_stop = False
                if (step + 1) % 5 == 0 or (step + 1) == epochs:
                    validation_data = dataset.get_validation_data(batch_size)
                    print("Doing evaluation on validation Set")
                    sys.stdout.flush()
                    early_stop = do_evaluation(session, eval_correct, validation_data, batch_size, images_placeholder, labels_placeholder)
		
                if (step + 1) == epochs or early_stop:
                    print("Doing evaluation on training set")
                    sys.stdout.flush()
                    do_evaluation(session, eval_correct, (training_data, training_labels), batch_size, images_placeholder, labels_placeholder)

                    print("Doing evaluation on the test set")
                    sys.stdout.flush()
                    test_data = dataset.get_test_data(batch_size)
                    do_evaluation(session, eval_correct, test_data, batch_size, images_placeholder, labels_placeholder)
                    
                    saver.save(session, "model.ckpt")

                    if (early_stop):
                        print("Achieved desired precision at step %d" % step)
                        return

run_training(89, 0.001, 40, 13)
